src/utils/simplifier.js

```diff
@@ -103,14 +103,53 @@ function simplify(realm, value: Value, isCondition: boolean = false): Value {
         // (x: boolean) || false <=> x
         if (!y.mightNotBeFalse()) return op === "||" ? x : realm.intrinsics.false;
       }
-      if (
-        op === "||" &&
-        y instanceof AbstractValue &&
-        y.kind === "||" &&
-        x.equals(y.args[0]) &&
-        !y.args[1].mightNotBeTrue()
-      )
-        return y;
+      if (op === "||") {
+        if (y instanceof AbstractValue && y.kind === "||") {
+          // x || x || y <=> x || y
+          if (x.equals(y.args[0])) return y;
+          // y || x || y <=> x || y
+          if (x instanceof AbstractValue && x.kind === "!") {
+            // !x || x || y <=> y
+            if (x.args[0].equals(y.args[0])) return y.args[1];
+            // !x || y || x <=> y
+            if (x.args[0].equals(y.args[1])) return y.args[0];
+          }
+        }
+      }
+      if (op === "&&") {
+        if (x instanceof AbstractValue) {
+          if (x.kind === "&&") {
+            // (x && y) && x <=> x && y
+            // (x && y) && y <=> x && y
+            if (x.args[0].equals(y) || x.args[1].equals(y)) return x;
+          } else if (x.kind === "!") {
+            // !x && x <=> false
+            if (x.args[0].equals(y)) return realm.intrinsics.false;
+          }
+        }
+        if (y instanceof AbstractValue && y.kind === "&&") {
+          // x && (x && y) <=> x && y
+          // y && (x && y) <=> x && y
+          if (x.equals(y.args[0]) || x.equals(y.args[1])) return y;
+        }
+        // x && (x && y || x && z) <=> x && (y && z)
```

Comment:
The comment seems wrong, while the implementation may be correct.

Reply:
The comment tries to communicate the pattern in the most abstract way. The code does not follow the names in the comment when things get complicated.

Reply:
I think the comment should be `// x && (x && y || x && z) <=> x && (y || z)`